rust/src/csv/reader.rs

```diff
@@ -52,6 +53,127 @@ use crate::record_batch::RecordBatch;
+                    } else if INTEGER_RE.is_match(v) {
+                        possibilities.push(4);
+                    } else {
+                        // we have a string, so the schema becomes a string
+                        return Field::new(&field_name, DataType::Utf8, false)
+                    }
+                }
+            },
+            None => {
+                // do nothing, we'll return an optional String
+                possibilities.push(0);
+            }
+        }
+    }
+    possibilities.sort();
+    possibilities.dedup();
+    if possibilities.len() == 1 {
+        match possibilities[0] {
+            0 => { return Field::new(&field_name, DataType::Utf8, has_nulls) },
+            1 => { return Field::new(&field_name, DataType::Boolean, has_nulls) },
+            2 => { return Field::new(&field_name, DataType::Int64, has_nulls) },
+            3 => { return Field::new(&field_name, DataType::Float64, has_nulls) },
+            4 => { return Field::new(&field_name, DataType::Int32, has_nulls) },
+            _ => { return Field::new(&field_name, DataType::Utf8, has_nulls) },
+        }
+    } else if possibilities.len() > 2 {
+        return Field::new(&field_name, DataType::Utf8, has_nulls)
+    } else {
+        if possibilities.contains(&1) || possibilities.contains(&2) {
+            // boolean or timestamp is incompatible with other types, return string
+            return Field::new(&field_name, DataType::Utf8, has_nulls)
+        } else if possibilities.contains(&3) && possibilities.contains(&4) {
+            // we have an integer and double, fall down to double
+            return Field::new(&field_name, DataType::Float64, has_nulls)
+        } else {
+            return Field::new(&field_name, DataType::Utf8, has_nulls)
+        }
+    }
+}
+
+/// Infer the schema of a csv file
+fn infer_schema(mut file: File, delimiter: u8) -> Schema {
+
+    file.seek(SeekFrom::Start(0)).unwrap();
+
+    let mut csv_reader = csv::ReaderBuilder::new()
+        .delimiter(delimiter)
+        .from_reader(BufReader::new(file.try_clone().unwrap()));
+
+    // get header names
+    let headers = &csv_reader.headers().unwrap();
+
+    let my_headers: Vec<&str> = headers.clone().iter().collect();
+    let header_length = my_headers.len();
+    let mut my_vals: Vec<Vec<Option<String>>> = vec![vec![]; header_length];
+
+    // return file seek back to start
+    file.seek(SeekFrom::Start(0)).unwrap();
+
+    let csv_reader = csv::ReaderBuilder::new()
+        .delimiter(delimiter)
+        .from_reader(BufReader::new(file.try_clone().unwrap()));
+
+    // take at most 100 rows for schema inference
+    // [TODO] it's often better to read the whole file and infer its schema
+    for result in csv_reader.into_records().take(100) {
+        let record = result.unwrap();
+
+        for i in 0..header_length {
+            let borrowed = record.clone();
+            my_vals[i].push(borrowed.get(i).map(|s| s.to_string()));
```

Comment:
Building an in-memory list of all values seems expensive. Could we just build a set of possible types per column here instead? That way the max number of items per column is very low (limited to the number of data types)

Reply:
Done, memory usage is much better, especially for low cardinality values. I tested with a 2GB csv file, and memory usage was reasonable, although inference is slow if I din't limit max records.

Reply:
I think I did a poor job explaining this... currently you build a HashSet of unique values per column (potentially pretty large) in one pass and then process that hashset to get a hashset of possible types ... what I was suggesting is why not just build the hashset of possible types as you go and then there is no need to store the actual values at all ... does that make sense?